source/bigint.c

```diff
@@ -0,0 +1,353 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/bigint.h>
+
+#define BASE_BITS 32
+#define NIBBLES_PER_DIGIT ((BASE_BITS) / 4)
+#define LOWER_32_BIT_MASK 0xFFFFFFFF
+#define INT64_MIN_AS_HEX 0x8000000000000000
+
+/*
+ * Working set of invariants:
+ *
+ * (1) Negative zero is illegal
+ * (2) The only time leading (trailing) 0-value digits are allowed is a single instance to represent actual zero.
+ *
+ * Other proposals:
+ *
+ * (1) Functionally immutable API (no visible side-affects)
+ * (2) params may be temporarily altered during computation before being restored in order to prevent pointless copying
+ *   (most often around sign negation and operator switching)
+ *
+ * Dev ideas:
+ *
+ * (1) Internal arithmetic ops may use a "raw" interface to allow for subsequence operations used in multiply/divide.
+ */
+
+void aws_bigint_clean_up(struct aws_bigint *bigint) {
+    aws_array_list_clean_up(&bigint->digits);
+}
+
+int aws_bigint_init(struct aws_bigint *bigint, struct aws_allocator *allocator) {
+    return aws_bigint_init_reserve(bigint, allocator, 0);
+}
+
+int aws_bigint_init_reserve(struct aws_bigint *bigint, struct aws_allocator *allocator, uint64_t reserve_bits) {
+    uint64_t digit_count = reserve_bits == 0 ? 1 : (reserve_bits - 1) / BASE_BITS + 1;
+
+    if (aws_array_list_init_dynamic(&bigint->digits, allocator, digit_count, sizeof(uint32_t))) {
+        return AWS_OP_ERR;
+    }
+
+    uint32_t zero = 0;
+    aws_array_list_push_back(&bigint->digits, &zero);
+
+    bigint->sign = 1;
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_advance_cursor_to_hex_start(struct aws_byte_cursor *hex_cursor) {
+    if (hex_cursor->len >= 2) {
+        const char *raw_ptr = (char *)hex_cursor->ptr;
+        if (raw_ptr[0] == '0' && (raw_ptr[1] == 'x' || raw_ptr[1] == 'X')) {
+            aws_byte_cursor_advance(hex_cursor, 2);
+        }
+
+        while (hex_cursor->len > 0 && *hex_cursor->ptr == '0') {
+            aws_byte_cursor_advance(hex_cursor, 1);
+        }
+    } else if (hex_cursor->len == 1) {
+        if (*hex_cursor->ptr == '0') {
+            aws_byte_cursor_advance(hex_cursor, 1);
+        }
+    }
+}
+
+static int s_uint32_from_hex(struct aws_byte_cursor digit_cursor, uint32_t *output_value) {
+    AWS_FATAL_ASSERT(digit_cursor.len <= NIBBLES_PER_DIGIT);
+
+    *output_value = 0;
+
+    while (digit_cursor.len > 0) {
+        char hex_digit = *digit_cursor.ptr;
+        uint32_t hex_value = 0;
+        if (hex_digit <= '9' && hex_digit >= '0') {
+            hex_value = hex_digit - '0';
+        } else if (hex_digit <= 'f' && hex_digit >= 'a') {
+            hex_value = hex_digit - 'a' + 10;
+        } else if (hex_digit <= 'F' && hex_digit >= 'A') {
+            hex_value = hex_digit - 'A' + 10;
```

Comment:
Not sure if we care about the efficiency of this fn, or if it's just for testing, but I did a [crazy table-lookup](https://github.com/awslabs/aws-c-http/blob/31b5dd632e5b27c5d538bb7b1604322cdc6da6db/source/strutil.c#L17) for the `int aws_strutil_read_unsigned_hex(struct aws_byte_cursor cursor, uint64_t *dst)` function I built for aws-c-http

feel free to copy the table, or maybe we can move this fn to common, or this is just for testing and whooooo caaares

Reply:
I'd like to reuse it, but want to hold off for now on moving things across libraries